radix-engine-stores/src/hash_tree/mod.rs

```diff
@@ -20,55 +23,289 @@ mod test;
+    changes: Vec<SubstateHashChange>,
 ) -> Hash {
-    let value_set: Vec<(Hash, Option<(Hash, SubstateId)>)> = changes
-        .iter()
-        .map(|(id, value_hash)| {
-            (
-                hash(scrypto_encode(id).unwrap()),
-                value_hash.map(|value_hash| (value_hash, id.clone())),
-            )
+    let changes_by_re_node_module = index_by_re_node_module(changes);
+    let mut nested_root_changes = Vec::new();
+    for (re_node_module, substate_changes) in changes_by_re_node_module {
+        let nested_root =
+            put_substate_changes(store, current_version, &re_node_module, substate_changes);
+        nested_root_changes.push(IdChange::new(re_node_module, nested_root));
+    }
+    put_changes(
+        store,
+        current_version,
+        current_version.unwrap_or(0) + 1,
+        nested_root_changes
+            .into_iter()
+            .map(|change| to_re_node_change(change))
+            .collect(),
+    )
+}
+
+// only internals below
+
+#[derive(
+    Debug, Clone, Copy, PartialEq, Eq, Hash, ScryptoCategorize, ScryptoEncode, ScryptoDecode,
+)]
+struct ReNodeModule {
+    re_node_id: RENodeId,
+    node_module_id: NodeModuleId,
+}
+
+impl ReNodeModule {
+    fn new(re_node_id: RENodeId, node_module_id: NodeModuleId) -> Self {
+        Self {
+            re_node_id,
+            node_module_id,
+        }
+    }
+}
+
+fn index_by_re_node_module(
+    changes: Vec<SubstateHashChange>,
+) -> HashMap<ReNodeModule, Vec<IdChange<SubstateOffset, Hash>>> {
+    let mut by_re_node_module = HashMap::<ReNodeModule, Vec<IdChange<SubstateOffset, Hash>>>::new();
+    for change in changes {
+        let substate_id = change.id;
+        by_re_node_module
+            .entry(ReNodeModule::new(substate_id.0, substate_id.1))
+            .or_insert_with(|| Vec::new())
+            .push(IdChange::new(substate_id.2, change.changed));
+    }
+    by_re_node_module
+}
+
+struct TreeRoot<P> {
+    hash: Hash,
+    node: TreeNode<P>,
+}
+
+fn put_substate_changes<S: TreeStore>(
+    store: &mut S,
+    current_version: Option<Version>,
+    re_node_module: &ReNodeModule,
+    changes: Vec<IdChange<SubstateOffset, Hash>>,
+) -> Option<TreeRoot<SubstateOffset>> {
+    let substate_entry = get_substate_entry(store, current_version, re_node_module);
+    let (substate_version, substate_root) = break_tuple(substate_entry);
+    let mut substate_tree_store = NestedTreeStore::new(store, re_node_module, substate_root);
+    let substate_root_hash = put_changes(
+        &mut substate_tree_store,
+        substate_version,
+        current_version.unwrap_or(0) + 1,
+        changes
+            .into_iter()
+            .map(|change| to_substate_change(change))
+            .collect(),
+    );
+    let substate_root_node = substate_tree_store
+        .extract_new_root()
+        .map_payload(|payload| match payload {
+            Payload::ReNodeModule(_) => panic!("unexpected payload in substate layer"),
```

Comment:
I'm a little confused about why we have this `Payload` enum? In the "top" tree, I imagine we just store `ReNodeModuleRoot`s, and in the bottom tree, I imagine we just store `Substate`s - so could this type not be fixed at compile time (as eg a generic parameter of the ReNodeModuleTreeStore / SubstateSubtreeStore)

Reply:
Or do we store both types in the same tree?

Reply:
Ahh wait - I guess this is because we wish to serialize both types in the same underlying node DB family?

Potentially an enum is still not required, and we can just know which leaf we're expecting from which tree we're in?

But I guess having an enum is useful anyway in case we want to change the payload in the future. So I don't feel super strongly 👍

Reply:
> Potentially an enum is still not required, and we can just know which leaf we're expecting from which tree we're in?

But how to do it ("expect") without an enum?

Reply:
See example PR: https://github.com/radixdlt/radixdlt-scrypto/pull/835/files - feel free to merge it in if you like it